Feature define the logging interface

LucasRoesler opened the pull request:
## Description
This creates a new logs endpoint that proxies the http request to the log provider as described in openfaas/faas#751


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

See openfaas/faas#751
And openfaas/faas#1130

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

This has been manually tested as well as some initial unit tests have been added to verify that it does not leak any goroutines.  I have used this in the faas-netes and faas-swarm projects

This can be tested using these instructions 
- https://gist.github.com/LucasRoesler/a6d0c5e2e669825ec4ab613e3487f1e8 (installs with the updated k8s provider)
 - https://gist.github.com/LucasRoesler/ec0c365d02296c29e66be683247be8aa

alexellis commented on httputils/writers.go:
```diff
@@ -0,0 +1,13 @@
+
+// WriteError sets the response status code and write formats the provided message as the
+// response body
+func WriteError(w http.ResponseWriter, statusCode int, msg string, args ...interface{}) {
```
Can we use `http.Error` instead because it's a part of stdlib?

LucasRoesler replied:
yeah, i am not sure why i added this here, i might have had a plan to do something with it that never happened

alexellis replied:
👍 Do you think the code in the proxy could change to `http.Error` too?

LucasRoesler replied:
I see now why the utility was original written, the internal utility wraps the string formatting, which is not supported by `http.Error`

```go
func Error(w ResponseWriter, error string, code int)
```

How do you feel about 

```go
// ErrorF sets the response status code and write formats the provided message as the
// response body
func ErrorF(w http.ResponseWriter, statusCode int, msg string, args ...interface{}) {
	http.Error(w, fmt.Sprintf(msg, args...), statusCode)
}
```

LucasRoesler replied:
I have replaced the utility with the `ErrorF` described above

alexellis commented on logs/logs.go:
```diff
@@ -0,0 +1,53 @@
+	// Name is the function name and is required
+	Name string `json:"name"`
+	// Instance is the optional container name, that allows you to request logs from a specific function instance
+	Instance string `json:"instance"`
```
Is this inspired/copied from anywhere in terms of naming / structure?

alexellis replied:
OK so Name -> FunctionName?

Instance -> the specific replica the function? i.e. get me the logs from replica 3/10?

What would be a valid identifier for "Instance" on Swarm vs Kubernetes?

LucasRoesler replied:
In swarm `Instance` == `Task.Name`
In k8s, `Instance` == `Pod.Name`

alexellis commented on logs/logs.go:
```diff
@@ -0,0 +1,53 @@
+	Follow bool `json:"follow"`
+	// Pattern is an optional regexp value to filter the log messages
+	Pattern *string `json:"pattern"`
+	// Invert allows you to control if the Pattern should be matched or negated
```
If that's the case wouldn't `InvertPattern` make more sense to first-time-readers?

LucasRoesler replied:
This is renamed

alexellis commented on logs/logs.go:
```diff
@@ -0,0 +1,53 @@
+	// Follow is allows the user to request a stream of logs
+	Follow bool `json:"follow"`
+	// Pattern is an optional regexp value to filter the log messages
+	Pattern *string `json:"pattern"`
```
I didn't think this would be in scope for v1. Did this feature creep in from the interfaces in Kubernetes/Swarm?

LucasRoesler replied:
A couple of reasons, 
1. I wanted to ensure that it could be implemented, 
2. it didn't take very long, 
3. and to ensure consistent logging behavior i wanted to emphasize it being done in the provider instead of trying to pass filters to the various providers in some way

alexellis commented on logs/handler.go:
```diff
@@ -0,0 +1,431 @@
+
+// Requestor submits queries the logging system.
+// This will be passed to the log handler constructor.
+type Requestor interface {
```
I hear `requester` is more common in spoken English. https://english.stackexchange.com/questions/29254/whats-the-difference-between-requester-and-requestor

LucasRoesler replied:
I have renamed it

alexellis commented on logs/handler.go:
```diff
@@ -0,0 +1,431 @@
+	"github.com/openfaas/faas-provider/httputils"
+)
+
+var upgrader = websocket.Upgrader{} // use default options
```
What if this was initialised closer to where it was used instead of being in a broader scope?

LucasRoesler replied:
I have removed the websocket implementation

alexellis commented:
Hi Lucas,

Thanks for running with this feature.

Can you give a brief summary on the work you've done here in addition to the commit messages?

Did you go for websockets? How do providers need to implement their logging code? What example / documentation do we need in README.md?

Alex

LucasRoesler commented:
@alexellis  I did try for websockets, as seen in the `NewWSLogHandlerFunc`.  It implements a version of the same code but using websockets instead of plain HTTP. It suffered from some of the same complexity issues related to timeouts but was infact harder for me to deal with resetting the timeouts (i am not sure i ever really resolved the timeouts issue _and_ closing everything). It also is much harder to test on a standard *nix system.  The plain HTTP  handler can be tested using curl without any special web knowledge .  Testing the websocket with something like curl requires awkward instructions. 

It is very easy to test the HTTP handler with curl and  reading the http chunk-encoded new-line delimited http stream was really easy in Go.  

Reading the stream in javascript is also pretty simple, for example using https://www.npmjs.com/package/request

```js
request.on('response', function (response) {
  response.on('data', function (chunk) {
     doSometing(JSON.parse(chunk))
  });
  response.on('end', function () {
    // done()
  });
});
```

If you are willing to try ReadableStreams, then it looks even nicer https://deanhume.com/experimenting-with-the-streams-api/

For new providers, we can use my PRs for Swarm and k8s as examples.  Additionally, I believe the absolute minimal example of a log server implementation would look like this

```go
package main

import (
	"context"
	"net/http"
	"time"

	"github.com/openfaas/faas-provider/logs"
)

// staticLogRequestor implements the logs Requestor returning a static stream of logs
type staticLogRequestor struct {
	logs []string
}

func (s staticLogRequestor) Query(ctx context.Context, r logs.Request) (<-chan logs.Message, error) {
	resp := make(chan logs.Message, len(s.logs))

	// A real implementation would possibly run a query to their log storage here, if it returns a
	// channel, and pass that channel to the go routine below instead of ranging over `s.logs`
	// If the log storage backend client does not return a channel, the query would need to
	// occur at the beginning of the goroutine below ... or a separate goroutine

	go func() {
		for _, m := range s.logs {
			// always watch the ctx to timeout/cancel/finish
			if ctx.Err() != nil {
				return
			}

			resp <- logs.Message{
				Name:      r.Name,
				Instance:  "fake",
				Timestamp: time.Now(),
				Text:      m,
			}
		}
	}()

	return resp, nil
}

func main() {
	requestor := staticLogRequestor{logs: []string{"msg1", "msg2", "something interesting"}}
	http.HandleFunc("/system/logs", logs.NewSimpleLogHandlerFunc(requestor))
	http.ListenAndServe(":80", nil)
}
```

alexellis commented:
One of the core use-cases for logs may be on OpenFaaS Cloud where it's impossible to run `kubectl` as a regular use without faas-cli or kubectl access. This may also be added to the Core OpenFaaS UI.

I would see a panel or page that would render either all the logs available or the last N lines statically, without a tailing streaming option. Is that possible with this implementation?

alexellis commented on logs/handler.go:
```diff
@@ -0,0 +1,431 @@
+}
+
+// NewSimpleLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewSimpleLogHandlerFunc(requestor Requestor) http.HandlerFunc {
```
@stefanprodan This is what code looks like when WriteTimeout is *disabled* on the http.Server.

LucasRoesler commented:
@alexellis i think i have addressed all of the comments.  I have tested this in faas-swarm along with the most recent updates to https://github.com/openfaas/faas/pull/1130 

I have not updated the server code to remove the WriteTimeout  until there is a final decision on https://github.com/openfaas/faas/issues/1147  Once that is decided I will remove the alternative implementations.

alexellis commented on logs/logs.go:
```diff
@@ -0,0 +1,58 @@
+	// Since is the optional datetime value to start the logs from
+	Since *time.Time `json:"since"`
+	// Limit sets the maximum number of log messages to return, <=0 means unlimited
+	Limit int `json:"limit"`
```
`Limit` -> `Tail`

alexellis commented:
Please can you include an example of bootstrapping in README.md for platform developers? I.e. @acornies @ewilde @sargun

LucasRoesler commented:
@alexellis i have pushed all of the changes per the comments and our conversation

alexellis commented:
This PR will have a squash, but I've requested two other people who haven't been as close to the code as we have to take a look over it before the final button press.

LucasRoesler commented:
@alexellis i have squashed everything down to the original commit.

alexellis commented on httputils/writers.go:
```diff
@@ -0,0 +1,12 @@
+package httputils
```
stdlib naming of `httptest` and `io/util` would direct us along the lines of using `httputil` - and what am I missing with `ErrorF`, wouldn't it be `Errorf`?

LucasRoesler replied:
updated

alexellis commented on httputil/writers.go:
```diff
@@ -0,0 +1,12 @@
+
+// Errorf sets the response status code and write formats the provided message as the
+// response body
+func Errorf(w http.ResponseWriter, statusCode int, msg string, args ...interface{}) {
```
@LucasRoesler thank you for changing this.

alexellis commented on proxy/proxy.go:
```diff
@@ -112,21 +112,21 @@ func proxyRequest(w http.ResponseWriter, originalReq *http.Request, proxyClient
 	functionName := pathVars["name"]
 	if functionName == "" {
-		writeError(w, http.StatusBadRequest, errMissingFunctionName)
+		httputils.ErrorF(w, http.StatusBadRequest, errMissingFunctionName)
```
This hasn't been changed to match the new name,  so I'm not sure how CI is passing? @LucasRoesler

LucasRoesler commented:
@alexellis i fixed the typos

LucasRoesler commented:
@johnmccabe and @stefanprodan this is now ready for your review

alexellis commented:
I'm sorry that neither of the people pinged to review this PR did that for you.

Let's just merge it and unblock you, I know you've been waiting a long time.

Alex
